### PR TITLE
Support files for Flatpak builds

### PR DESCRIPTION
**Pull request by aperezdc**
:warning: **~~This is a working proof of concept, but I want to share it to request feedback~~** :warning: 

In particular, I would like to iron out at least the following:

- [x] Application ID. For now it's `im.riot.Riot`, but maybe something else is more appropriate (comments welcome).
   - Changed to `im.vector.Riot` as per Matthew's comment.
- [x] Cleaning up files used for build which are unneeded at runtime. There are options which can be added to the JSON manifest for this, I just didn't have time to add them and test.
- [x] AppStream metadata is missing. This would be needed for the application to appear e.g. in GNOME Software.
  - [ ] It would be nice to have a screenshot of Riot available at some public URL to link it in the AppStream metadata.
- [ ] Adding Flatpak builds to CI.
- [ ] Figuring out with the Matrix developers in which way a Flatpak repository could be made available online for people to use. Probably needs some discussion first.

---

Install the dependency base Electron bundle with:

```sh
flatpak --user remote-add endless-electron-apps --from \
    https://s3-us-west-2.amazonaws.com/electron-flatpak.endlessm.com/endless-electron-apps.flatpakrepo
flatpak --user install endless-electron-apps io.atom.electron.BaseApp
```

then Riot can be built and saved to a Flatpak repository with (it takes a bit, it will build NodeJS inside the Flatpak sandbox):

```sh
flatpak-builder --repo=flatpak-repo flaptak-build \
    electron/build/im.riot.Riot.json
```

and installed locally for testing with:

```sh
flatpak --user remote-add --no-gpg-verify \
    riot-local file://`pwd`/flatpak-repo
flatpak --user install riot-local im.riot.Riot
flatpak run im.riot.Riot
```

### CODE REVIEW

**Comment by aperezdc** [comment]
One thought about the application identifier: probably it should be `org.matrix.Riot`, or even maybe better `im.vector.Riot` given that the GitHub organization is `vector-im`. I guess we want @ara4n or someone else from the core team to vet one of those.

**Comment by aperezdc** [comment]
I have added AppStream metadata:

* The `<description>` is filled with a couple of paragraphs taken from [riot.im](https://riot.im) — let me know if some other text is preferred.
* Screenshots are missing. For adding this we would be nice to host the image somewhere, to be able to write the URL in the XML file.  This should be in 16:9 ratio, with a minimum width of 620px (more info [in the spec](https://www.freedesktop.org/software/appstream/docs/chap-Quickstart.html#qsr-app-screenshots-info)).

**Comment by ara4n** [comment]
the application ID should definitely be im.vector.Riot - Riot's not made by Matrix.org but Vector.  (Matrix.org is the non-profit; Vector is a separate startup building apps & services on top of Matrix).

This is looking really good. We can provide some 'official' screenshots and check the description field once you're ready for the PR to be reviewed and landed :)

**Comment by aperezdc** [comment]
I have just added a few changes:

- The application ID is not `im.vector.Riot` (thanks @ara4n for the clarification).
- Allow using `/tmp` from the Flatpak sandbox, which is needed for Electron's unique instance checks.
- Allow sending D-Bus messages to the notification daemon from gthe sandbox, which makes desktop notifications work.

**Comment by ara4n** [comment]
@aperezdc is there anything to be done on this other than screenshots from us?

**Comment by aperezdc** [comment]
@ara4n From my hitlist above, there are only two things missing: Making CI do Flatpak builds, and making a Flatpak repository available. Probably we can just merge this PR, and skip these two for now.

---

I think that figuring out how to provide a Flatpak repository for people to use can be done separately and shouldn't keep this from being merged. Probably it needs some discussion before and it may take a while to set up something.

As for the CI, would it make sense to make Travis-CI do also Flatpak builds? I suppose it's fine to skip this, but it would be nice to have. So far I haven't had some free time to sit down to figure out how to get `flatpak-builder` to work inside Travis (it could be as well that it doesn't work because of the union mounts and the setuid sandbox helper). I think it's fine to merge this PR and I can make a separate follow-up PR which modifies `.travis.yml` later on (if I ever get it to work).

**Comment by Croydon** [comment]
You might want to have a look at Flathub. However I'm not sure if this is already ready for real usage.

https://github.com/flathub/flathub/wiki
https://flathub.org/builds/

**Comment by TingPing** [comment]
> You might want to have a look at Flathub. However I'm not sure if this is already ready for real usage.

Packager for Flathub here. So the main blocker at a glance is you currently pass `--share=network` as a build option and this is not acceptable on the Flathub build servers. I'm not super familiar with `npm` but I did find a project called `npm-bundle` that might be useful, if releases are made using this it might work without network access.

**Comment by TingPing** [comment]
As for Travis CI, that won't be easy because Travis is far too ancient for Flatpak and running `flatpak-builder` inside a `docker` container or such is PITA.

File: electron/build/im.vector.Riot.desktop [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[Desktop Entry]
+Name=Riot
+Comment=A feature-rich client for Matrix.org
+Exec=riot-web
+Terminal=false
+Type=Application
+Icon=im.vector.Riot
+StartupWMClass="Riot"
+Categories=Network;InstantMessaging;Chat;IRCClient
```
Comment (t3chguy) [inline electron/build/im.vector.Riot.desktop]: This doesn't match the cats being used in the Linux app, (though Chat should be in them and I missed it)
i.e remove IRCClient

**Comment by aperezdc** [comment]
I have added a couple of commits to move the files under `electron_app/`, fix the categories in the `.desktop` file, and update the used sources to Riot version 0.10.2

**Comment by dbkr** [comment]
This looks nice, although I'm struggling to find any kind of docs on what the various options in the manifest are and the --share options seems to make no appearance in the flatpak-builder man page (I'm assuming it allows network access for the build process so npm can download the rest of the deps?)

Main thing here is wondering if this would be better in its own repo? ie. the same layout as a build hosted on flathub (see https://github.com/flathub/org.blender.Blender or indeed https://github.com/flathub/im.riot.Riot)

If that sounds good, I can pull all this out into its own repo.

File: electron_app/build/im.vector.Riot.json [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    "base": "io.atom.electron.BaseApp",
+    "base-version": "master",
+    "runtime": "org.freedesktop.Platform",
+    "runtime-version": "1.4",
+    "sdk": "org.freedesktop.Sdk",
+    "command": "riot-web",
+    "finish-args": [
+        "--device=dri",
+        "--filesystem=home",
+        "--filesystem=/tmp",
```
Comment (TingPing) [inline electron_app/build/im.vector.Riot.json]: You should set `$TMPDIR` at runtime to `$XDG_CACHE_HOME`

**Comment by TingPing** [comment]
> I'm struggling to find any kind of docs on what the various options in the manifest are and the --share options seems to make no appearance in the flatpak-builder man page

`man flatpak-build-finish`

File: electron_app/build/im.vector.Riot.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            "sources": [
+                {
+                    "type": "archive",
+                    "url": "https://github.com/vector-im/riot-web/archive/v0.12.2.tar.gz",
+                    "sha256": "d31713f6efcd07366bf0d3b7364f547504c958a35d07a37bed809ef71789e2bf"
+                },
+                {
+                    "type": "file",
+                    "path": "flatpak-Makefile",
+                    "dest-filename": "Makefile"
```
Comment (TingPing) [inline electron_app/build/im.vector.Riot.json]: In general these are avoided these days. You can do:

```json
{
  "name": "riot-web",
  "buildsystem": "simple",
  "build-commands": [
    "install -Dm644 foo /app/share/foo",
    "etc"
  ]
}

**Comment by dbkr** [comment]
@TingPing thanks! I had completely failed to realise `flatpak-build-finish` was separate from `flatpak-builder`.

**Comment by TingPing** [comment]
@dbkr `flatpak-builder` is a tool built *on top* of `flatpak build` so there is some overlap in where docs discuss some things.

**Comment by jonathancross** [comment]
Is anyone still working on this PR?

**Comment by dbkr** [comment]
I think this is superseded by https://github.com/vector-im/element-desktop/pull/293

**Pull request closed without merging**